aptos-move/aptos-intent/src/codegen.rs

```diff
@@ -0,0 +1,489 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    ArgumentOperation, BatchArgument, BatchedFunctionCall, PreviousResult, APTOS_INTENT_KEY,
+};
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{PartialVMError, PartialVMResult},
+    file_format::*,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, TypeTag},
+    metadata::Metadata,
+    transaction_argument::TransactionArgument,
+    vm_status::StatusCode,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Default)]
+struct Context {
+    address_pool: BTreeMap<AccountAddress, AddressIdentifierIndex>,
+    identifier_pool: BTreeMap<Identifier, IdentifierIndex>,
+    module_pool: BTreeMap<(AddressIdentifierIndex, IdentifierIndex), ModuleHandleIndex>,
+    function_pool: BTreeMap<(ModuleHandleIndex, IdentifierIndex), FunctionHandleIndex>,
+    struct_pool: BTreeMap<(ModuleHandleIndex, IdentifierIndex), StructHandleIndex>,
+    signature_pool: BTreeMap<Signature, SignatureIndex>,
+    script: CompiledScript,
```

Comment:
If timing permits, I'd love to see this evolving into a builder that supports both `CompiledScript` and `CompiledModule`. This doesn't mean you need to add all the import functions though -- those can be added on-demand.